src/OrchardCore.Modules/OrchardCore.Resources/Assets/monaco/ocmonaco-editor.js

```diff
@@ -0,0 +1,154 @@
+                linkedElement.value = editor.getValue();
+            });
+
+            // debounced change handler that will trigger the preview
+            const debouncedOnChange = debounce(function(event) {
+                linkedElement.value = editor.getValue();
+                $document.trigger('contentpreview:render');
+            }, 250);
+            // trigger preview when content changes (debounced)
+            editor.getModel().onDidChangeContent(debouncedOnChange);
+
+            // dynamically update the height of the monaco editor to a max of 650px 
+            const updateHeight = () => {
+                let width = el.parentElement.clientWidth;
+                const contentHeight = Math.min(650, editor.getContentHeight());
+                el.style.width = `${width}px`;
+                el.style.height = `${contentHeight + 20}px`;
+                editor.layout({ width: width - 20,  height: contentHeight });
+            };
+            editor.onDidContentSizeChange(updateHeight);
+            updateHeight();
+
+            // dynamically set the width when the browser is resized.
+            const debouncedUpdateHeight = debounce(updateHeight,150);
+            window.addEventListener('resize', debouncedUpdateHeight);
+            
+            // include the shortcodes action for html editors
+            if(el.dataset.shortcodes)
+            {
+                const shortcodesAction = {
+                    id: "shortcodes",
+                    label: "Add Shortcode",
+                    run: function (editor) {
+                        shortcodesApp.init(function (value) {
+                            if (value) {
+                                var selection = editor.getSelection();
+                                var text = value;
+                                var op = { range: selection, text: text, forceMoveMarkers: true };
+                                editor.executeEdits("shortcodes", [op]);
+                            }
+                            editor.focus();
+                        })
+                    },
+                    contextMenuGroupId: 'orchardcore',
+                    contextMenuOrder: 0,
+                    keybindings: [
+                        monaco.KeyMod.Alt | monaco.KeyCode.KEY_S, 
+                    ]
+                }
+                editor.addAction(shortcodesAction);
+            }
+
+            // set the jQuery monaco data to the dom node. 
+            // The data object is used to allow other components to use the editor functions or updateHeight function
+            $el.data("monaco", { editor, updateHeight });
+        }
+
+        // loop through and initialize all monaco editors on the page
+        document.querySelectorAll("[data-monaco-oc]").forEach(initMonacoEditor);
+
+        // initialize the monaco editor when a widget is dynamically added to the page
+        document.addEventListener('widget-added', function (e) {
```

Comment:
I think this event should be generalized to "initialize-editors" and triggerred whenever the editors need to be initialized. Any action that would require an editor to be "initialized" or "updated" should trigger the event. This way, we won't have to handle all cases (such as widget shown, hidden, or tabs opened)